span_compressed.go

```diff
@@ -0,0 +1,321 @@
+		}
+		return true
+	}
+	// If the span has not parent and no transaction, it should be reported by
+	// the caller.
+	if s.tx == nil {
+		return false
+	}
+
+	if !s.isCompressionEligible() {
+		// At this point, the span isn't compressable which is likely to be a
+		// parent or non-compressable sibling, either way, the transaction or
+		// the span's buffer needs to be flushed and `false` is returned.
+		if !s.tx.buffer.empty() {
+			s.tx.buffer.flush(nil)
+		}
+		if !s.buffer.empty() {
+			s.buffer.flush(nil)
+		}
+		return false
+	}
+
+	// The span is compressable and we need to store in the parent's cache.
+	if s.tx.buffer.empty() {
+		s.tx.buffer.store(s)
+		return true
+	}
+
+	// When the span is compressable, try to compress it, report back true:
+	// On success: store it.
+	// On failure: flush and swap the cache with the current span.
+	if s.tx.buffer.span.compress(s) {
+		s.tx.buffer.span.buffer.store(s)
+	} else {
+		s.tx.buffer.flush(s)
+	}
+	return true
+}
+
+func (s *Span) isCompressionEligible() bool {
+	if s == nil {
+		return false
+	}
+	return s.exit && !s.ctxPropagated &&
+		(s.Outcome == "" || s.Outcome == "success")
+}
+
+// A span is eligible for compression if all the following conditions are met
+// 1. It's an exit span
+// 2. The trace context has not been propagated to a downstream service
+// 3. If the span has outcome (i.e., outcome is present and it's not null) then
+//    it should be success. It means spans with outcome indicating an issue of
+//    potential interest should not be compressed.
+// The second condition is important so that we don't remove (compress) a span
+// that may be the parent of a downstream service. This would orphan the sub-
+// graph started by the downstream service and cause it to not appear in the
+// waterfall view.
+func (s *Span) compress(sibling *Span) bool {
+	strategy := s.canCompressComposite(sibling)
+	if strategy == 0 {
+		strategy = s.canCompressStandard(sibling)
+	}
+
+	// If the span cannot be compressed using any strategy.
+	if strategy == 0 {
+		return false
+	}
+
+	if s.composite.empty() {
+		s.composite = compositeSpan{
+			count:               1,
+			sum:                 int64(s.Duration),
+			compressionStrategy: strategy,
+		}
+	}
+
+	s.composite.count++
+	s.composite.sum += int64(sibling.Duration)
+	siblingTimestamp := sibling.timestamp.Add(sibling.Duration)
+	if siblingTimestamp.After(s.composite.lastSiblingEndTime) {
+		s.composite.lastSiblingEndTime = siblingTimestamp
+	}
+	return true
+}
+
+func (s *Span) canCompressStandard(sibling *Span) int {
+	if !s.isSameKind(sibling) {
+		return 0
+	}
+
+	if s.isExactMatch(sibling) {
+		if s.durationLowerOrEq(sibling,
+			s.tx.compressedSpans.exactMatchMaxDuration,
+		) {
+			return compressedStrategyExactMatch
+		}
+		return 0
+	}
+	if s.isSameKind(sibling) {
+		if s.durationLowerOrEq(sibling,
+			s.tx.compressedSpans.sameKindMaxDuration,
+		) {
+			s.Name = "Calls to " + s.Context.destinationService.Resource
```

Comment:
can this be deferred to the flush, so we just do it once for a set of compressed spans, and not for each one? (it'll incur an allocation)

Reply:
yeah I had a note there to optimize but I must've deleted it, I'll change where we do this.